activerecord/lib/active_record/associations/has_one_association.rb

```diff
@@ -61,7 +61,7 @@ def replace(record, save = true)
             save &&= owner.persisted?
 
             transaction_if(save) do
-              remove_target!(options[:dependent]) if target && !target.destroyed? && assigning_another_record
+              remove_target!(options[:dependent]) if target && !target.destroyed? && assigning_another_record && owner.persisted?
```

Comment:
I think this extra check should be done on line 60.
If we don't want to save anything, we shouldn't be in the `transaction_if` block.

Reply:
> I think this extra check should be done on line 60.
> If we don't want to save anything, we shouldn't be in the `transaction_if` block.

I agree, I would prefer to put the conditional here.

Reply:
I was trying this approach, but we want to be in the `transaction_if(save)` even though we don't want to save things. For example `set_inverse_instance(record)` doesn't always mean saving the record. Right? A possible solution might be adding 
```ruby
  set_inverse_instance(record) if record && assigning_another_record
```